internal/goroutine/recorder/recorder.go

```diff
@@ -0,0 +1,234 @@
+package recorder
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/sourcegraph/log"
+	"github.com/sourcegraph/sourcegraph/internal/rcache"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"k8s.io/utils/strings/slices"
+)
+
+type Recordable interface {
+	Start()
+	Stop()
+	Name() string
+	Type() RoutineType
+	JobName() string
+	SetJobName(string)
+	Description() string
+	Interval() time.Duration
+	RegisterRecorder(recorder *Recorder)
+}
+
+type Recorder struct {
+	rcache      *rcache.Cache
+	logger      log.Logger
+	recordables []Recordable
+	hostName    string
+}
+
+// seenTimeout is the maximum time we allow no activity for each host, job, and routine.
+// After this time, we consider them non-existent.
+const seenTimeout = 6 * 24 * time.Hour // 6 days
+
+const keyPrefix = "background-job-logger"
+
+// maxRecentRunsLength is the maximum number of recent runs we want to store for each routine.
+const maxRecentRunsLength = 100
+
+// New creates a new recorder.
+func New(logger log.Logger, hostName string, cache *rcache.Cache) *Recorder {
+	return &Recorder{rcache: cache, logger: logger, hostName: hostName}
+}
+
+// Register registers a new routine with the recorder.
+func (m *Recorder) Register(r Recordable) {
+	m.recordables = append(m.recordables, r)
+}
+
+// RegistrationDone should be called after all recordables have been registered.
+// It saves the known job names, host names, and routine names in Redis, along with updating their “last seen” date/time.
+func (m *Recorder) RegistrationDone() {
+	// Save/update known job names
+	for _, jobName := range m.collectAllJobNames() {
+		m.saveKnownJobName(jobName)
+	}
+
+	// Save known host name
+	m.saveKnownHostName()
+
+	// Save/update all known recordables
+	for _, r := range m.recordables {
+		m.saveKnownRoutine(r)
+	}
+}
+
+// collectAllJobNames collects all known job names in Redis.
+func (m *Recorder) collectAllJobNames() []string {
+	var allJobNames []string
+	for _, r := range m.recordables {
+		jobName := r.JobName()
+		if slices.Contains(allJobNames, jobName) {
+			continue
+		}
+		allJobNames = append(allJobNames, jobName)
+	}
+
+	return allJobNames
```

Comment:
```suggestion
	names := make(map[string]struct{}, len(m.recordables))
	for _, r := range m.recordables {
		names[r.JobName()] = struct{}{}
	}
	allJobNames := make([]string, 0, len(names))
	for name := range names {
		allJobNames = append(allJobNames, name)
	}
	return allJobNames
```

Slightly more idiomatic - checks for duplicates with two scans instead of a nested loop.